### ISSUE TITLE
Errors when getting album covers of large music collection

### ISSUE THREAD

**Issue by n00bundercover** [owncloud/music#653]
On my nextcloud instance I got a user owning ~20k audio files.
When I log in as this user and start the Music app, I always get these error messages:

```
Error | PHP | fread()  expects parameter 1 to be resource, null given at  /usr/share/webapps/nextcloud/apps2/music/3rdparty/getID3/getid3/getid3.php#1723
Error | PHP | ftell()  expects parameter 1 to be resource, null given at  /usr/share/webapps/nextcloud/apps2/music/3rdparty/getID3/getid3/getid3.php#1700
Error | PHP | fseek()  expects parameter 1 to be resource, null given at  /usr/share/webapps/nextcloud/apps2/music/3rdparty/getID3/getid3/getid3.php#1758
Error | PHP | Undefined index: filesize at /usr/share/webapps/nextcloud/apps2/music/3rdparty/getID3/getid3/module.tag.apetag.php#25
Error | PHP | fread()  expects parameter 1 to be resource, null given at  /usr/share/webapps/nextcloud/apps2/music/3rdparty/getID3/getid3/getid3.php#1723
Error | PHP | ftell()  expects parameter 1 to be resource, null given at  /usr/share/webapps/nextcloud/apps2/music/3rdparty/getID3/getid3/getid3.php#1700
Error | PHP | fseek()  expects parameter 1 to be resource, null given at  /usr/share/webapps/nextcloud/apps2/music/3rdparty/getID3/getid3/getid3.php#1758
Error | PHP | Undefined index: filesize at /usr/share/webapps/nextcloud/apps2/music/3rdparty/getID3/getid3/module.tag.apetag.php#25
```
When I do the same thing with a user who just has a couple of audio files (<10), I do not get any errors.
Unfortunately, I do not really know which files cause the issue, or if its maybe just related to the big number of audio files.

I am using:

arch
nginx
postgresql
php 7.2
nextcloud 13 stable channel
Music app 0.7.0

As always: Thanks for the awesome app!

**Comment by paulijar** [comment]
The app should be able to handle ~20k tracks without problems; I have been testing it with >58k tracks.

My best guess is that these log lines come when the app tries to read embedded album art from an audio file. Can you maybe see two albums with no cover art once the UI has finished loading?

It might be that the internal book-keeping of Nextcloud is not up-to-date with the actual disk contents: some files have been deleted from the disk but Nextcloud still thinks that they exist. If this is the case, you should be able to fix it by running `sudo -u www-data ./occ files:scan <username>`. After that, you may want to run also `sudo -u www-data ./occ music:cleanup`; this removes any unavailable files from the records of the Music app (the same cleanup task is run also automatically in the background task once per 15 minutes).

**Comment by n00bundercover** [comment]
OK I'll stop mentioning the library size in future... I felt special! ;-)

Yes, there are issues with cover art; not just two missing, but it's hard to explain, so here is a video of it:
https://peertube.video/videos/watch/0f771e47-4702-48ab-aa56-279a05682b36
(Note: It wont always stop loading at the same point it did in the video, it varies. But it will never fully load. Also often cover art further down the list will be already loaded, while there is a huge gap not showing anything above it)

Will try to run the commands later when I got access to the server.
I know it's a vague bug description, I just don't know how to get further info!

**Comment by paulijar** [comment]
> OK I'll stop mentioning the library size in future... I felt special! ;-)

I have to admit that my personal music collection isn't nearly that large. The 58k+ tracks is something I just fabricated to be able to test with a large collection :D. But the largest collection mentioned in any bug report was ~150k tracks which was way more than the Music app could handle; the app broke down already during scanning.

> But it will never fully load. 

So the cover art further down the list are still missing, say, after 30 minutes? That's sound quite odd, if you only get those few error lines to you log. It's normal that it may take ~1 second per cover to load the first time you open the music collection, but eventually the covers should appear. And on the next time you open the app, the loading should be faster as the covers are already cached on the server side. And on the third and further times it should be extra fast as the covers get cached also on the client side.

I also noticed one weird thing on your video, there was a play icon shown in front of each track. That is no supposed to happen, the icon should be visible only on hover and while a track is playing. You should try reloading the page on browser. On Chrome, you need to use ctrl+reload to make sure that the browser reloads all the .css and .js files; on Firefox that should happen on reload by default.

**Comment by n00bundercover** [comment]
OK, so I tried both commands two posts, both said 0 files changed.

Yes, cover art is missing no matter how long I leave the tab open. Tried Firefox and Chromium based browser (brave) now, but both got the same issue. Also, both show the play arrows also when not hovering over a track! Could it be related to the dark theme?

To clarify: I wanted to keep things simple and reproducable, so these errors in log are from only starting the music app! I opened the music app as the user in question, would not scroll or play a track or anything like that, but switch to admin user instantly and have a look at the log. I am pretty sure that once I do more stuff, more error logs appear. Will try to get some more useful logs this evening.

Last but not least: Do you think this could be an error in my cache settings, rather than a bug in the music app?

**Comment by paulijar** [comment]
> Also, both show the play arrows also when not hovering over a track! Could it be related to the dark theme?

Yes, you are right. I just gave a try for the nextcloud-breeze-dark theme and it does cause this issue. The theme seems to specifically override some css rules of the Music app. But I guess that these override rules are designed for some older version of Music and are not fully compatible with the most recent versions.

> Last but not least: Do you think this could be an error in my cache settings, rather than a bug in the music app?

Difficult to say, but it may be part of the problem. One theory would be that in your setup the system now gets overwhelmed by the amount of files tried to be read simultaneously. I guess that you have thousands of albums in total and most of them have album art set? That would mean that thousands of files need to be opened for reading and some layer could run out of resources if no one throttles the operation.

Do you see anything interesting in `/var/log/php7.2-fpm.log` or `/var/log/nginx/error.log`?

**Comment by paulijar** [comment]
This is probably related to the issue  #719.

**Comment by paulijar** [comment]
@n00bundercover I just released v0.12.0 where album covers are lazy-loaded instead of loading them all at once. Could you test if this solves your reported issue?

**Comment by paulijar** [comment]
I assume that lazy-loading the album-covers has solved this issue. Please reopen if the problem still persists on Music v0.13.0 (or later).
